crates/bevy_app/src/app.rs

```diff
@@ -862,16 +865,42 @@ fn run_once(mut app: App) {
     app.cleanup();
 
     app.update();
+
+    let mut exit_code_reader = ManualEventReader::default();
+    if let Some(app_exit_events) = app.world().get_resource::<Events<AppExit>>() {
+        if exit_code_reader
+            .read(app_exit_events)
+            .any(|exit| *exit == AppExit::Error)
+        {
+            return AppExit::Error;
+        }
+    }
+
+    AppExit::Success
 }
 
-/// An event that indicates the [`App`] should exit. If one or more of these are present at the
-/// end of an update, the [runner](App::set_runner) will end and ([maybe](App::run)) return
-/// control to the caller.
+/// An event that indicates the [`App`] should exit. If one or more of these are present at the end of an update,
+/// the [runner](App::set_runner) will end and ([maybe](App::run)) return control to the caller.
 ///
 /// This event can be used to detect when an exit is requested. Make sure that systems listening
 /// for this event run before the current update ends.
-#[derive(Event, Debug, Clone, Default)]
-pub struct AppExit;
+#[derive(Event, Debug, Clone, Default, PartialEq, Eq)]
+pub enum AppExit {
+    /// [`App`] exited without any problems.
+    #[default]
+    Success,
+    /// The [`App`] experienced an unhandleable error.
+    Error,
```

Comment:
could you make this variant hold a status code, and return it?

Reply:
We could. I just don't think it's a good idea. Status codes aren't very standardized e.g. anything above 255 might get trucated on linux (see [`process:exit` platform-specific behaviour](https://doc.rust-lang.org/std/process/struct.ExitCode.html#differences-from-processexit)). The std's solution is [`ExitCode`](https://doc.rust-lang.org/std/process/struct.ExitCode.html#differences-from-processexit) but I don't think letting people produce a `AppExit::Error(ExitCode::Success)` makes much sense.

Reply:
then a `NonZeroU8`?